### ISSUE TITLE
Add a failover if error occurs when push data

### ISSUE THREAD

**Issue by remy56k** [wavelog/WaveLogGate#23]
Hi! First, thanks for WaveLogGate & Wavelog, awesome stack.

Is it possible to add a failover behavor if Internet connection fail or timeout occurs ?
- https://github.com/wavelog/WaveLogGate/blob/875e996cb49f368b97db4f76010ba05a66fd2402/main.js#L200
- https://github.com/wavelog/WaveLogGate/blob/875e996cb49f368b97db4f76010ba05a66fd2402/main.js#L207

My idea is simple : write the QSO in a local ADIF file if error, so it'll be easy to import the file when back home. 

I can try to implement this but I'm not very used to compiling JS for cross-platform, I use/prefer Golang for this kind of interfaces.

Thanks,

**Comment by int2001** [comment]
At first i liked the idea. But the longer i'm thinking about it, questions are coming up:
Why would one save an ADIF twice as failover?

i mean: If you are doing FT8 with WJST-X, the WSJT-X creates his own ADIF regardless if reporting to UDP (for WLGate) is set nor not. You can simply upload the WSJT-X-ADIF to Wavelog. With it's automatic dupe-check only the new ones are going to be added to Wavelog.

Or i am wrong here?

**Comment by remy56k** [comment]
Thanks for consideration !

Indeed, but : 

- WSJT ADIF log can be huge
- WSJT ADIF log is more difficult to find on the different OS than a path provided by the user
- If the WLG is hidden behind the WSJT window, you don't see if an error occurs, if just some QSO fail (4G sharing, bad networking), you cannot notice there are missing entries in your log. 
- I know WaleLog can handle with duplicate entries, but with a 10Mo WSJT file for just 5 missing QSO, it seems not very handy.

New idea : maybe just ask for a folder instead a file path, and make error files like : `YYYYMMDD_missing-QSO.adi`.
If a file exists with the date of the day, some QSO are missing and need to be imported.

**Comment by int2001** [comment]
Risk that those files are ignored by OP is the same as at WSJT-X.
So redundant Data is growing and growing.

**Comment by remy56k** [comment]
I agree, but the case of a down WaveLog server or instable network are exceptions, and user knows where the folder is, the file or folder can be purged after import.
WSJT logs can not be purged without loose done countries. 

Maybe add the feature as optional with a check box to enable it if needed ?
I think it's important in /P operations to be sure QSO are not lost when using WLG.

Thanks for consideration,

**Comment by oe3gwu** [comment]
> Hi! First, thanks for WaveLogGate & Wavelog, awesome stack.
> 
> Is it possible to add a failover behavor if Internet connection fail or timeout occurs ?
> 
>     * https://github.com/wavelog/WaveLogGate/blob/875e996cb49f368b97db4f76010ba05a66fd2402/main.js#L200
> 
>     * https://github.com/wavelog/WaveLogGate/blob/875e996cb49f368b97db4f76010ba05a66fd2402/main.js#L207
> 
> 
> My idea is simple : write the QSO in a local ADIF file if error, so it'll be easy to import the file when back home.
> 
> I can try to implement this but I'm not very used to compiling JS for cross-platform, I use/prefer Golang for this kind of interfaces.
> 
> Thanks,

Why not simply add a retry button for another push? No need for an ADIF.

**Comment by int2001** [comment]
If we add a "Retry"-Button, we need to implement some kind of queuing into it.
And: The User needs to press that button. For every QSO? For the queue? What happens if 2 QSOs are failing, the next 2 are not failing, and afterwards everything works again?
What happens if the User is "in the field", autostarts WavelogGate and makes QSOs without a internet-connection? Every QSO would fail in that case.

WavelogGate is a simple "transporter" for ADIFs. Never meant to be a buffering/queuing/dequeuing system.

the paradigm for WLGate was (and is): Keep it as simple as possible.

**Comment by oe3gwu** [comment]
> If we add a "Retry"-Button, we need to implement some kind of queuing into it. And: The User needs to press that button. For every QSO? For the queue? What happens if 2 QSOs are failing, the next 2 are not failing, and afterwards everything works again? What happens if the User is "in the field", autostarts WavelogGate and makes QSOs without a internet-connection? Every QSO would fail in that case.
> 
> WavelogGate is a simple "transporter" for ADIFs. Never meant to be a buffering/queuing/dequeuing system.
> 
> the paradigm for WLGate was (and is): Keep it as simple as possible.

Why do you think so complicated?

EVERY variant you mentioned has the same problem now. When you are not connected to the internet, everything fails. My problem is, that out of 100 QSO's 1 fails to upload because of timeout. If this happens, a simple retry button would be enough. No need for buffering, because when it is failed, it is still buffered, because it is still on the screen, just red and not green.

**Comment by oe3gwu** [comment]
> @oe3gwu - feel free to implement it. but please consider "simpleness" as well and failover-mechanisms. It needs to be bulletproof. Otherwise there will be support-questions over and over.
> 
> that's one of the benefits OpenSource provides. You can fork, implement it and give something back via a PR.
> 
> No offense meant ;) Just showing a way how to achieve it.
> 
> [Edit because reacting to "y do you think so overcomplicated]: Software-development is only 10% real coding. it's 90% brainwork. Checking side-effects, checking support-loads and so on.

Honestly if I would be a programmer, I would do it myself. But I am not!

**Comment by int2001** [comment]
And I am doing these things during my spare time as a hobby :) 
Furthermore I offered a simple solution: upload the ADIF in case of network-problems.
